ENH: Slightly improve error when gufunc axes has wrong size

seberg opened the pull request:
My hope was to tweak it into something useful that:

    a @= b

can raise when `b` should have two dimensions and has two axes specified but actually only has one.
I didn't succeed, but I still think it a slight improvement to give the ufunc name and the actual core dimensions.

seberg committed:
DOC: Slightly improve error when gufunc axes has wrong size

My hope was to tweak it into something useful that:

    a @= b

can raise when `b` should have two dimensions and has two axes specified
but actually only has one.
I didn't succeed, but I still think it a slight improvement to give the
ufunc name and the actual core dimensions.

mhvk commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -1847,8 +1843,11 @@ _parse_axes_arg(PyUFuncObject *ufunc, int op_core_num_dims[], PyObject *axes,
+            PyErr_Format(PyExc_TypeError,
+                    "%s: operand %d has %d core dimensions, "
+                    "but the axes item is not tuple.  (A single integer is "
+                    "also allowed when only one dimension is required.)",
```
I think I'd leave out the parenthetical sentence here, or write it as "A single integer is only allowed when just one dimension is required" -- but I think it does not necessarily help, and people should just read the documentation if confused.

seberg replied:
Yeah, agreed. Removed.  Something doesn't quite feel ideal still, but not sure what :), and I think its an improvement either way.

mhvk reviewed (approved):
Yes, good to be more explicit!  One super nitpicky comment; can ignore if you wish.

seberg commented:
@mhvk one thing I noticed looking here, is that if core axes are optional, you cannot omit them by passing a shorter axes tuple.  I.e. `matul(2D, 2D, axes=[(-2, -1)], (-1,), (-1))` doesn't work.  Just curious if it is intentional, or because it would be complicated to allow?

mhvk commented:
@seberg - I think I introduced the ability to give axes just before we allowed for optional axes, so I think it is more a question of oversight...

It might actually be quite neat if one could indeed use `axes` to signal one really wanted `matrix-stack @ vector-stack` rather than a stack of matrices. Though perhaps `(mat @ vec[..., np.newaxis]).squeeze(-1)` or just `np.einsum(...)` ends up being clearer. And the alternative of just implementing `matvec` and `vecmat` is perhaps cleanest of all.

mhvk commented:
p.s. This seems all OK, so fine to merge.

mattip commented:
Could these be `AxisError` instead of `ValueError` ?

seberg commented:
Hmmm, right now we only use `AxisError` for an out-of-bounds axis I think, not for a different way `axis=` can be invalid.

So yeah, makes sense, although for myself I am happy to not broaden up `AxisError` for this (happy to do so I am just ±0 on it).

mattip commented:
Something other than ValueError would be helpful in order to properly catch errors in ` a @= b`

seberg committed:
ENH: Use AxisError when gufunc axes appear wrong due to the number of entries

This allows catching the error relatively targeted for in-place matmul `a @= b`
which may use this path.

seberg commented:
Heh, I forgot about that idea...  I guess that is a good enough reason here, we might want to do it for more axis/axes related `ValueError`'s I guess.

mattip commented:
This [is failing](https://github.com/numpy/numpy/actions/runs/3564647967/jobs/5988864709#step:4:3350). It used to raise `TypeError("axes item 2 should be a tuple")`.
```
a = np.arange(27.).reshape((3, 3, 3))
b = np.arange(10., 19.).reshape((3, 1, 3))
assert_raises(TypeError, inner1d, a, b, axes=[-1, -1, -1])
```

seberg committed:
MAINT: Restore most TypeErrors (a bit more compexl than nice, but...)

seberg commented:
Grrr, restored the `TypeError` where it clearly is correct the long way around...  Not pretty, but maybe the practical path?

mhvk reviewed (approved):
Better logic indeed, even if a bit more tortuous.

mattip committed:
DOC: add a release note

mattip reviewed (approved):
LGTM. I added a release note, please take a look before merging.
